src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs

```diff
@@ -1108,8 +1109,33 @@ private static uint HandleEventPeerRecvAborted(State state, ref StreamEvent evt)
 
         private static uint HandleEventStartComplete(State state, ref StreamEvent evt)
         {
-            // Store the start status code and check it when propagating shutdown event, which we'll get since we set SHUTDOWN_ON_FAIL in StreamStart.
-            state.StartStatus = evt.Data.StartComplete.Status;
+            uint status = evt.Data.StartComplete.Status;
+
+            // The way we expose Open(Uni|Bi)directionalStreamAsync operations is that the stream
+            // is also accepted by the peer (i.e. it is within advertised stream limits). However,
+            // We may receive START_COMPLETE notification before the stream is accepted, so we defer
+            // completing the StartcompletionSource until we get PeerAccepted notification.
+
+            if (status != MsQuicStatusCodes.Success)
+            {
+                // Start irrecoverably failed. The possible status codes are:
+                //   - Aborted - connection aborted by peer
+                //   - InvalidState - stream already started before, or connection aborted locally
+                //   - StreamLimitReached - only if QUIC_STREAM_START_FLAG_FAIL_BLOCKED was specified (not in our case).
+                //
+                // We disregard duplicate StreamStart calls
+                Debug.Assert(status == MsQuicStatusCodes.Aborted || status == MsQuicStatusCodes.InvalidState);
+                state.StartCompletionSource.SetException(
+                    ExceptionDispatchInfo.SetCurrentStackTrace(GetConnectionAbortedException(state)));
```

Comment:
I believe we should preserve the logic to separate "aborted" from other errors. What I mean is throw ConnectionAbortedException only on Aborted, and generic QuicException on InvalidState and others

https://github.com/dotnet/runtime/blob/42044c65e9a09ae3d6f056d53bb1af83aa6ae449/src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs#L181-L187

Reply:
Currently, we throw `ConnectionAbortedException` or `QuicOperationAbortedException` depending on whether we have a connection abort code (that we would have already received due to the order of callbacks). While not ideal, I believe this is a better solution than the alternative.

The problem with throwing `QuicException` on `InvalidState` is that that there is a data race we cannot get easily rid of: if one thread closes the connection locally, and another thread tries to open a stream, it may happen that the stream open call will happen after the call to `CloseAsync`, but before we receive event from MsQuic.

Introducing a guard on the managed side and checking it during Open*StreamAsync does not really help, as the threads can still be scheduled in a way that hits the invalid state (e.g. thread 1 checks guard, thread 2 CloseAsync, thread1 calls StreamStart).

Last, but not least, if StreamStart is scheduled before ConnectionClose, we may complete the TCS with `QuicOperationAbortedException`, as we do elsewhere, which would mean that the code may throw `QuicOperationAbortedException` or `QuicException` just based on thread scheduling and user potentially needs to watch out for both.

Reply:
INVALID_STATE is an indicator that what user is doing is wrong; and while sometimes the user might get away with something wrong because they lucked out (e.g. using non-thread-safe object in multithreaded environment might sometimes work) I still think we should preserve the information when we have it.

We had a similar discussion about preserving the exception for the code I linked https://github.com/dotnet/runtime/pull/67342#discussion_r838850848

We might choose later to incorporate the two exceptions together (as an inner exception or as an additional field), but the point is that user should have an access to the information in one way or another.

Reply:
I have tried to change this to
```cs
                if (status == MsQuicStatusCodes.Aborted)
                {
                    state.StartCompletionSource.SetException(
                        ExceptionDispatchInfo.SetCurrentStackTrace(GetConnectionAbortedException(state)));
                }
                else
                {
                    state.StartCompletionSource.SetException(
                        ExceptionDispatchInfo.SetCurrentStackTrace(new QuicException($"StreamStart finished with status {MsQuicStatusCodes.GetError(status)}")));
                }
```

but then tests started mysteriously failing:
```
 Failed System.Net.Quic.Tests.MsQuicTests.OpenStreamAsync_ConnectionAbort_Throws(unidirectional: True, localAbort: True) [14 ms]
  Error Message:
   Assert.Throws() Failure
Expected: typeof(System.Net.Quic.QuicOperationAbortedException)
Actual:   typeof(System.Net.Quic.QuicException): StreamStart finished with status INVALID_STATE
---- System.Net.Quic.QuicException : StreamStart finished with status INVALID_STATE
  Stack Trace:
     at System.Net.Quic.Implementations.MsQuic.MsQuicStream.HandleEventStartComplete(State state, StreamEvent& evt) in C:\source\runtime\src\libraries\System.Net.Quic\src\System\Net\Quic\Implementations\MsQuic\MsQuicStream.cs:line 1134    
   at System.Net.Quic.Implementations.MsQuic.MsQuicStream.NativeCallbackHandler(IntPtr stream, IntPtr context, StreamEvent* streamEvent) in C:\source\runtime\src\libraries\System.Net.Quic\src\System\Net\Quic\Implementations\MsQuic\MsQuicStream.cs:line 919
```

relevant part of the test:
```cs

            // Open one stream, second call should block
            QuicStream stream = await OpenStreamAsync(clientConnection);
            ValueTask<QuicStream> waitTask = OpenStreamAsync(clientConnection);
            Assert.False(waitTask.IsCompleted);

            if (localAbort)
            {
                await clientConnection.CloseAsync(0);
                await Assert.ThrowsAsync<QuicOperationAbortedException>(() => waitTask.AsTask().WaitAsync(TimeSpan.FromSeconds(3)));
            }
```

so it looks like we can hit this race even when we don't run multithreaded.

Reply:
Does this happen sometimes or on every run?

Reply:
I tried only a few times but it was reliable

Reply:
This is not a connection aborted by the peer but ourselves. So `INVALID_STATE` here is expected, user is doing something wrong. I'm rather surprised we get sometimes `ABORTED`.

Reply:
> This is not a connection aborted by the peer but ourselves. So INVALID_STATE here is expected,

Because the connection is aborted by us, then I as a user would expect all pending operations to throw `QuicOperationAbortedException`, in line with https://github.com/dotnet/runtime/issues/55619 (as currently outlined).

> user is doing something wrong.

If by the user you mean the test code above, then I disagree. The connection is accessed from a single thread, and opening the stream was attempted before trying to close the connection. I can imagine that in a real use case, there would be more logic between the call to OpenStreamAsync and CloseAsync (behind some condition). There is no data race evident in the user code. If we start throwing `QuicException` with `INVALID_STATE`, we start leaking implementation details.

> I'm rather surprised we get sometimes ABORTED.

Side note: this can while we are waiting for stream limit to increase (i.e. PEER_ACCEPTED event) and peer closes the connection on us.